### PR TITLE
Add SSM and s3 endpoints for private subnets.

### PR DESCRIPTION
**Pull request by dogfish182**
This PR adds 2 capabilities, 

Create an s3 endpoint on private subnets.
Create the neccessary endpoints for AWS session manager to control machines, these endpoints are 
    - ssm_endpoint
    - ec2messages_endpoint
    - ec2_endpoint
    - ssmmessages_endpoint

Our usecase for adding this PR was 

We wanted all of our resources in private subnets to only use S3 via VPC endpoints, we also wanted to access all of our machines using session manager and have all that traffic stay internal to the AWS network. To do this it requires creating the above listed interface endpoints.

### CODE REVIEW

File: endpoints.tf [hunk 1]
```diff
@@ -13,3 +13,147 @@ resource "aws_vpc_endpoint" "transfer_server" {
   vpc_endpoint_type   = "Interface"
   vpc_id              = aws_vpc.default.id
 }
+
+
+data "aws_region" "current" {}
```
Comment (shoekstra) [inline endpoints.tf:18]: This should be moved to main.tf as it's a shared resource, not specific to VPC endpoints.

File: endpoints.tf [hunk 2]
```diff
@@ -13,3 +13,147 @@ resource "aws_vpc_endpoint" "transfer_server" {
...
+resource "aws_vpc_endpoint" "ssmmessages_endpoint" {
+  count               = length(var.ssmmessages_endpoint.subnet_ids) != 0 ? 1 : 0
+  private_dns_enabled = var.ssmmessages_endpoint.private_dns_enabled
+  security_group_ids  = length(var.ssmmessages_endpoint.security_group_ids) == 0 ? [aws_security_group.sgp_ssmmessages_endpoint[count.index].id] : var.ssmmessages_endpoint.security_group_ids
+  service_name        = data.aws_vpc_endpoint_service.ssmmessages_endpoint[0].service_name
+  subnet_ids          = var.ssmmessages_endpoint.subnet_ids
+  tags                = merge(var.tags, { "Name" = "ssmmessages-endpoint-${var.name}" })
+  vpc_endpoint_type   = "Interface"
+  vpc_id              = aws_vpc.default.id
+}
```
Comment (shoekstra) [inline endpoints.tf:159]: A few changes needed here, please consider the following:

- Sort each resource's fields as per comment in `examples/all/main.tf` diff
- Avoid the resource type in the resource name, e.g.
  - `resource "aws_security_group" "sgp_ssmmessages_endpoint" ` vs `resource "aws_security_group" "ssmmessages_endpoint"`
  - `resource "aws_vpc_endpoint" "ssmmessages_endpoint"` vs `resource "aws_vpc_endpoint" "ssmmessages"`
- Consider separating each VPC endpoint with a comment for easier readability

File: endpoints.tf [hunk 3]
```diff
@@ -13,3 +13,147 @@ resource "aws_vpc_endpoint" "transfer_server" {
...
+  count             = var.private_s3_endpoint != false ? 1 : 0
+  service_name      = "com.amazonaws.${data.aws_region.current.name}.s3"
+  tags              = merge(var.tags, { "Name" = "s3-vpc-endpoint-${var.name}" })
+  vpc_endpoint_type = "Gateway"
+  vpc_id            = aws_vpc.default.id
+  route_table_ids   = aws_route_table.private[*].id
+}
+
+data "aws_vpc_endpoint_service" "ssm_endpoint" {
+  count   = length(var.ssm_endpoint.subnet_ids) != 0 ? 1 : 0
```
Comment (svanharmelen) [inline endpoints.tf:30]: As per the comment already made by @shoekstra, prefer `null` to indicate something should not be set/used. So this would then become:
```suggestion
  count   = var.ssm_endpoint != null ? 1 : 0
```

File: endpoints.tf [hunk 4]
```diff
@@ -13,3 +13,147 @@ resource "aws_vpc_endpoint" "transfer_server" {
...
+  service = "ssm"
+}
+
+resource "aws_security_group" "sgp_ssm_endpoint" {
```
Comment (svanharmelen) [inline endpoints.tf:34]: As these are all different resource types, they can all share the same name. So we don't have to prefix them with something the represents the type of resource:

```suggestion
resource "aws_security_group" "ssm_endpoint" {
```

File: endpoints.tf (hunk 3 above)
Comment (dogfish182) [reply endpoints.tf:30]: Ok, so I have the count on length because to use an ssm_endpoint you MUST give subnet ids (as its an interface, no interface in a subnet, no endpoint). Therefore I can do this 
```

variable "ssm_endpoint" {
  type = object({
    private_dns_enabled = bool
    security_group_ids  = list(string)
    subnet_ids          = list(string)
  })
  default = {
    private_dns_enabled = false,
    security_group_ids  = [],
    subnet_ids          = []
  }
  description = "Variables to provision an ssm endpoint to the VPC"
}
```

so the defaults are actually set.

The reason I'm setting defaults is that I want to enable two things 

1. choosing to deploy an endpoint or not (and those defaults will result in 'not')
2. but I also want to allow the ability to pass already created security groups to the module (endpoint creation can take more than one if you want). However, it's easier usually to not pass any security groups at which point the module will create them. 

Doing it like this is the only way I could make terraform happy, because I was trying to use 'null' and then 

`output "ssmmessages_endpoint_security_group_ids" {
  value = length(var.ssmmessages_endpoint.security_group_ids) == 0 ? aws_security_group.ssmmessages_endpoint.*.id : var.ssmmessages_endpoint.security_group_ids
}
`
the outputs would complain as they types based on the conditional output were not equivalent.

tl;dr 
I'm defaulting the values to not null, because I want conditional behaviour for the security groups and its the only way I could get it to work. 

Any pointers on how to fix that another way would be welcome as I fought with it for quite a while

File: outputs.tf [hunk 5]
```diff
@@ -77,3 +77,19 @@ output "private_route_table_ids" {
...
+  value = length(var.ec2messages_endpoint.security_group_ids) == 0 ? aws_security_group.sgp_ec2messages_endpoint.*.id : var.ec2messages_endpoint.security_group_ids
+}
+
+output "ec2_endpoint_security_group_ids" {
+  value = length(var.ec2_endpoint.security_group_ids) == 0 ? aws_security_group.sgp_ec2_endpoint.*.id : var.ec2_endpoint.security_group_ids
+}
+
+output "ssmmessages_endpoint_security_group_ids" {
+  value = length(var.ssmmessages_endpoint.security_group_ids) == 0 ? aws_security_group.sgp_ssmmessages_endpoint.*.id : var.ssmmessages_endpoint.security_group_ids
+}
```
[the remark this answers is not in the document]
Comment (shoekstra) [reply outputs.tf:95]: These outputs need to be sorted with other outputs 😉

File: .gitignore [hunk 6]
```diff
@@ -9,3 +9,6 @@
 *.tfvars
 
 **/crash.log
+
+# code editors
+.idea
```
Comment (shoekstra) [inline .gitignore:14]: This should go into your personal gitignore file to avoid bloat; add something like this to your ~/.gitconfig:

```
[core]
        excludesfile = /Users/shoekstra/.gitignore_global
```

And then add any files/dir to ignore, mine looks like this:

```
.env
tags
```

File: endpoints.tf (hunk 3 above)
Comment (shoekstra) [reply endpoints.tf:30]: Inputs to resources are also exported, so you could first test if your variable is set and either print an empty string or the value passed to the resource, something like this::

```hcl
output "ssmmessages_endpoint_security_group_ids" {
  value = var.ssmmessages_endpoint == null ? "" : aws_vpc_endpoint.ssmmessages_endpoint.*.security_group_ids
}
```

Let's discuss to work out best way forward.

**Review by svanharmelen — APPROVED** [review]
Made a few tiny tweaks (committed them directly to speed things up and get this on in), but if your good with this I'm good to merge it now...

**Review by shoekstra — APPROVED** [review]
